scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/SchedulingGoalDSLCompilationService.java

```diff
@@ -7,51 +7,75 @@
 
 import javax.json.Json;
 import javax.json.stream.JsonParsingException;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.io.OutputStreamWriter;
 import java.io.StringReader;
+import java.nio.charset.StandardCharsets;
 import java.util.List;
 
 import static gov.nasa.jpl.aerie.json.BasicParsers.productP;
 import static gov.nasa.jpl.aerie.json.BasicParsers.stringP;
 
 public class SchedulingGoalDSLCompilationService {
+
+  private final InputStream inputStream;
+  private final OutputStream outputStream;
+  private final InputStream errorStream;
+  private final Process nodeProcess;
+
+  public SchedulingGoalDSLCompilationService() throws SchedulingGoalDSLCompilationException, IOException {
+    final var schedulingDslCompilerPath = System.getenv("SCHEDULING_DSL_COMPILER_PATH");
+    final var commands = new String[] {"node", schedulingDslCompilerPath};
+    this.nodeProcess = Runtime.getRuntime().exec(commands);
+    try {
+      this.inputStream = this.nodeProcess.getInputStream();
+      this.outputStream = this.nodeProcess.getOutputStream();
+      this.errorStream = this.nodeProcess.getErrorStream();
+    } catch (Exception e) {
+      throw new SchedulingGoalDSLCompilationException("Could not create node subprocess: ", e);
+    }
+    System.out.println("SchedulingGoalDSLCompilerService created");
+  }
+
+  public void close() {
+    this.nodeProcess.destroy();
+  }
+
   private static final JsonParser<GoalDefinition> schedulingJsonP =
       productP
           .field("abc", stringP)
           .map(Iso.of(
               val -> new GoalDefinition("abc", val),
               GoalDefinition::value));
 
-  public static GoalDefinition compileSchedulingGoalDSL(final String goalTypescript) throws SchedulingGoalDSLCompilationException {
-    final var preface = "function goal(): {abc: string}\n";
-    final var epilogue = "\nJSON.stringify(goal())";
+  public GoalDefinition compileSchedulingGoalDSL(final String goalTypescript, final String goalName)
+  throws SchedulingGoalDSLCompilationException, IOException
+  {
+    var inputWriter = new OutputStreamWriter(this.outputStream, StandardCharsets.UTF_8);
+    var outputReader = new BufferedReader(new InputStreamReader(this.inputStream));
+    var errorReader = new BufferedReader(new InputStreamReader(this.errorStream));
 
-    final var executionResult = executeTypescript(preface, goalTypescript, epilogue);
-    if (!(executionResult instanceof final String executionResultString)) {
-      throw new SchedulingGoalDSLCompilationException("Expected result of evaluating typescript to be a string, but got " + executionResult);
+    inputWriter.write("{ \"source\": \"" + goalTypescript + "\", \"filename\": \"" + goalName + "\" }\n");
+
+    var error = errorReader.readLine();
+    var output = outputReader.readLine();
```

Comment:
I poked around at it this morning, and isolated the issue to these three lines.

First, I needed to add a `inputWriter.flush()` call after the write - the subprocess was never triggering its `on('data')` callback.

Next, there's an issue that I couldn't figure out how to fix without changing the protocol - both `readLine` calls are blocking, but only one of them is expected to actually have data (the subprocess with _either_ write to standard out, or to standard error).

The solution that worked for me was to change the protocol to make the subprocess always write to standard out, but it would always write two lines, the first of which was either "success" or "error".

Would you like me to commit these changes directly to your branch? Or merge this one into mine and keep working from there?

My only gut concern about the long-running subprocess approach is whether we'd ever have to restart it if it craps out for some reason. We can flesh that scenario out in some more testing.

Reply:
Let's merge this into yours and work from there. Do you have time for some pair programming on this today?